src/engine/QueryPlanner.cpp

```diff
@@ -175,58 +226,57 @@ std::vector<QueryPlanner::SubtreePlan> QueryPlanner::optimize(
         }
       }
 
-      // if our input is not optional, this means we still can arbitrarily
-      // optimize among our candidates and just append our new candidates.
-      if (!v[0]._isOptional) {
+      // all variables we have seen so far are considered "bound" from now on
+      // this also includes optionals. in the case of bind operations
+      // we do not have any variables yet, so this does no harm
+      {
         auto vc = v[0]._qet->getVariableColumns();
         std::for_each(vc.begin(), vc.end(), [&boundVariables](const auto& el) {
           boundVariables.insert(el.first);
         });
+      }
+
+      // if our input is not optional this means we still can arbitrarily
+      // optimize among our candidates and just append our new candidates.
+      if (!v[0]._isOptional) {
         candidatePlans.push_back(std::forward<decltype(v)>(v));
         return;
       }
 
-      // we are an optional, optimization across is forbidden.
-      // optimize all previously collected candidates, and then perform
-      // an optional join.
-      auto tg = createTripleGraph(&candidateTriples);
-      LOG(TRACE) << "Collapse text cliques..." << std::endl;
-      tg.collapseTextCliques();
-      LOG(TRACE) << "Collapse text cliques done." << std::endl;
-      // always apply all filters to be safe.
-      // TODO<joka921> it could be possible, to allow the DpTab to leave
-      // results unfiltered and add the filters later, but this has to be
-      // carefully checked and I currently see no benefit.
-      auto lastRow =
-          fillDpTab(tg, rootPattern->_filters, candidatePlans).back();
+      auto lastRow = optimizeCommutativ(candidateTriples, candidatePlans,
+                                        rootPattern->_filters);
       candidateTriples._whereClauseTriples.clear();
       candidatePlans.clear();
 
-      // join all the candidates plans for the complete previous inputs
-      // in an optional way with all the candidates for the contents of
-      // the optional join.
-      std::vector<SubtreePlan> optJoinCandidates;
-      for (const auto& a : lastRow) {
-        for (const auto& b : v) {
-          auto vec = createJoinCandidates(a, b, std::nullopt);
-          optJoinCandidates.insert(optJoinCandidates.end(),
-                                   std::make_move_iterator(vec.begin()),
-                                   std::make_move_iterator(vec.end()));
+      std::vector<SubtreePlan> nextCandidates;
+      if (v[0]._isOptional) {
+        // join all the candidates plans for the complete previous inputs
+        // in an optional way with all the candidates for the contents of
+        // the optional join.
```

Comment:
// For each candidate plan, and each plan from the OPTIONAL, create a new plan with an optional join. Note that createJoinCandidates will know that b is from an OPTIONAL.